### PR TITLE
Sound Classifier: allow user to control custom neural network structure

### LINKED ISSUE #1477: Sound Classifier Should Allow Tuning the Structure of the Custom Neural Network
Currently the custom neural network created by the sound classifier is always [two dense layer with 100 units each](https://github.com/apple/turicreate/blob/3236aa1e64ec76a4f8b775432ff46ba22b983115/src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py#L174). This is a sensible default but it should be configurable. The user should be able to specify how many dense layers they want and how many units should be in each layer.

### PR DESCRIPTION
**Pull request by TobyRoseman**
Fixes #1477

### CODE REVIEW

**Comment by nickjong** [comment]
For this kind of thing, I'm most worried about how we communicate about this API in the user guide. To what degree are we encouraging developers to spend a lot of time grid searching parameter values here? What benefit are they likely to see for their investment in time and energy? How should they think about these parameters?

**Comment by TobyRoseman** [comment]
> For this kind of thing, I'm most worried about how we communicate about this API in the user guide. To what degree are we encouraging developers to spend a lot of time grid searching parameter values here? What benefit are they likely to see for their investment in time and energy? How should they think about these parameters?

I agree this is something which will need to be carefully explained in our user guide. I created #1639 to make sure we explain this feature (as well as reusing deep feature and manually tweaking data augmentation) in our user guide. All three of theses features will be explained in an "Advanced Usage" section.

Here are my main thoughts about this parameter:
* I think for most users the default should work quite well. I see tuning this as more of an advanced features.
* If the user is training on a GPU and they are reusing deep features, then it should be very quick to try out a large number of different structures.
* If the number of classes is small, then they could use this feature to get a much smaller exported model with minimal affect on accuracy.

File: src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py [hunk 1]
```diff
@@ -135,6 +144,8 @@ def create(dataset, target, feature, max_iterations=10, verbose=True,
         raise _ToolkitError("'%s' column is not audio data." % feature)
     if target not in dataset.column_names():
         raise _ToolkitError("Target column '%s' does not exist" % target)
+    if not isinstance(custom_layer_sizes, list) or len(custom_layer_sizes) == 0 or not isinstance(custom_layer_sizes[0], int):
```
Comment (shantanuchhabra) [inline src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py]: Nit: What if someone passes in an array of the form [75, 200, None]? We probably want a check for all elements in the array?

We may want to add unit tests for corner cases (but maybe rarely encountered cases) like these?

File: src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py [hunk 2]
```diff
@@ -135,6 +144,8 @@ def create(dataset, target, feature, max_iterations=10, verbose=True,
...
+        raise _ToolkitError("'custom_layer_sizes' must be a non-empty list of integers." % target)
```
Comment (shantanuchhabra) [inline src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py]: I think "% target" is a typo?

File: src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py [hunk 3]
```diff
@@ -342,7 +362,14 @@ def _load_version(cls, state, version):
         # Load the custom nerual network
         num_classes = state['num_classes']
         num_inputs = state['_feature_extractor'].output_length
-        net = SoundClassifier._build_custom_neural_network(num_inputs, num_classes)
+        if 'custom_layer_sizes' in state:
+            # These are deserialized as floats
+            custom_layer_sizes = map(int, state['custom_layer_sizes'])
```
Comment (shantanuchhabra) [inline src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py]: Do we want to make it `list(map(...))` to not break on Python 3?

File: src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py (hunk 1 above)
Comment (TobyRoseman) [reply src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py]: Good points. I added more checks here and manually tested the cases you mentioned.

File: src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py (hunk 2 above)
Comment (TobyRoseman) [reply src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py]: Yeah, typo. Fixed.

File: src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py (hunk 3 above)
Comment (TobyRoseman) [reply src/unity/python/turicreate/toolkits/sound_classifier/sound_classifier.py]: Thanks. Done.
